porcelain/chain.go

```diff
@@ -27,3 +30,8 @@ func ChainBlockHeight(ctx context.Context, plumbing chPlumbing) (*types.BlockHei
 	}
 	return types.NewBlockHeight(currentHeight), nil
 }
+
+// SampleChainRandomness samples randomness from the chain at the given height.
+func SampleChainRandomness(ctx context.Context, plumbing chPlumbing, sampleHeight *types.BlockHeight) ([]byte, error) {
```

Comment:
BLOCKING: I want to make sure we are on the same page about the below before this gets merged.  I can be persuaded I am wrong but want to make the case.

For the purpose of deduplication and understandability I would prefer that we always sampled randomness using the same high level steps.  For good reasons we currently do the following steps in state machine randomness sampling:
(1) get the ancestors we need for all sampling calls through a channel and put them in a slice
(2) sample randomness by doing operations on the slice

In the current world (1) is done by `chain:GetRecentAncestors`, and (2) is the vm.Context `Rand` function.  Why do this instead of sampling randomness from the channel directly?  Well the state machine must in general be able to sample randomness from a variety of heights and doesn't have the ability to spin up new block history channels because chain depends on consensus right now and we can't have import cycles.  Copying a bunch of channels ahead of time could work but it is less clear *and* less efficient*.

The way `GetRecentAncestors` is written actually makes it a good candidate for step (1) here in this porcelain call, because no storage miner who is not past the point of being so late they can never submit a PoST will need more tipsets than those returned by `GetRecentAncestors`.  However it may be conceptually cleaner to offer this sampling function for the whole chain and just take all tipsets as step (1).  step (2) can be deduplicated in the same way it is now.

Without breaking things up explicitly into these steps [this PR does both steps together in one function](https://github.com/filecoin-project/go-filecoin/pull/2205/files#diff-5bfb40b2443579a645872ed2734bb760R755).
Because, thanks to @phritz, `GetRecentAncestors` was refactored modularly, you *can* still reuse a fair amount of the functionality of step (1) for the statemachine even when doing both steps at the same time (see my comments under this function).  However we'll miss out on the full benefit of just calling the same function in two places.

Finally if we keep the current arrangement of sampling randomness from a channel it feels like we are doing something wrong in the state machine case because we are taking tipsets out of a channel and putting them in a slice, only to take them from the slice and put them in a channel.  

*You can do some cool caching and lazy eval stuff to efficiently accomplish multiple calls with a single channel wrapped in a closure but then your design doc just gets roasted until you do the the clearer thing anyway.

Reply:
> Finally if we keep the current arrangement of sampling randomness from a channel it feels like we are doing something wrong in the state machine case because we are taking tipsets out of a channel and putting them in a slice, only to take them from the slice and put them in a channel.

The `Context.ancestors` slice's values are put into a channel so that we can share code with the storage miner. Assuming that the `Context.ancestors` slice looks more or less identical to `ChainLs` output (which it should, right?), we can easily use a channel to abstract over both sources of TipSet.